### ISSUE TITLE
Implications of update.packages

### ISSUE THREAD

**Issue by kmkostka** [ohdsi-studies/Covid19CharacterizationCharybdis#55]
This line of code is highly problematic:
````
# First, it probably is best to make sure you are up-to-date on all existing packages.
# Important: This code is best run in R, not RStudio, as RStudio may have some libraries
# (like 'rlang') in use.
#update.packages(ask = "graphics")
````

1. We are asking someone to update their entire package library. The majority, of which, are not utilized by the Charybdis Library. This can take a significant amount of time and, in my opinion, has little computational benefit. It would be better to have a more complete DESCRIPTION file with nested dependencies and/or CodeToRun that explicitly calls which libraries are really crucial to update. Some people have older versions of packages for specific compatibility reasons (e.g. [the notorious ggplot / ParallelLogger issue](https://github.com/OHDSI/ParallelLogger/issues/5)). This seems like a user's choice not something we should be putting as a mandate in a package.

2. The mysterious 'rlang' reference has no context. Most people are going to ignore it. It seems like the reference is trying to allude to 'rlang' issues with API thresholds. Presumably, by ignoring it, a user may get an error like this:
````
Error: Failed to install 'unknown package' from GitHub:
  Failed to install 'unknown package' from GitHub:
  Failed to install 'unknown package' from GitHub:
  HTTP error 403.
  API rate limit exceeded for 34.207.230.121. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)

  Rate limit remaining: 0/60
  Rate limit reset at: 2020-06-27 04:33:36 UTC

  To increase your GitHub API rate limit
  - Use `usethis::browse_github_pat()` to create a Personal Access Token.
  - Use `usethis::edit_r_environ()` and add the token as `GITHUB_PAT`.
Error: `call` is deprecated as of rlang 0.3.0
````

This issue is more of a FYI that this documentation should not be replicated in a future study.

**Comment by schuemie** [comment]
Wouldn't you always want to be up-to-date with all your packages? For example, the ggplot2 - ParallelLogger issue was resolved a while back, being up-to-date would therefore avoid that problem. I mentioned 'rlang' because a lot of people have found they are not able to update that specific package while in RStudio. 

I highly recommend using `renv` to handle package dependencies. As described [here](https://forums.ohdsi.org/t/using-renv-to-handle-r-package-dependencies/10914), it has some nice properties:

- Each study package can have its own R library. So different studies can be active simultaneously, each with different dependencies, and you can quickly switch between them without the need to reinstall the dependencies.

- An R library can be rebuilt from scratch based on a so-called ‘lock file’. This file describes all the packages that must be installed, which precise versions, and where to install them from.

- Most of the packages are installed from binaries, not from source, thus avoiding the issues I observed with packrat. It is also quite fast (rebuilding an environment using packrat could take many hours).  

I would highly recommend using the [function I created in OhdsiRTools](http://ohdsi.github.io/OhdsiRTools/reference/createRenvLockFile.html) to create the lock file, and then distributing the lock file to allow people to install the study package and its dependencies, as described in the forum post mentioned above. Let me know if you're interested, and we can work through it together.

**Comment by kmkostka** [comment]
There isn’t a legitimate use case here to update every single library. It’s a waste of time during set up to ask someone to update every single library. It should not be a requirement for this study. 

The problem with versioning is people may be in the middle of projects that have code that utilizes older libraries. Even if a package issue is resolved (e.g. ggplot - ParallelLogger), it doesn’t mean someone can update the whole kit. The introduction of Andromeda and sunset of FFbase is a great example. I cannot install the newer versions of packages for my old PLE studies without retooling the PLE design and doing unit testing to validate. So if I had to replicate a prior study and were rerunning the analysis package as it was intended you may opt to run on the same version of the library you used when you ran the study the first time. 

````renv```` works for a lot of environments but it is unlikely to be the only approach. We got plenty of pushback on the OHDSI Forum of people who anticipated local environment issues. It’s the same kind of pushback for Docker. The reality is there’s no one size fits all here.

The best practice should be an enumerated list (preferably in a computable format) of the minimum version necessary for all dependent libraries. The user can decide which install option is most appropriate given their local environment constraints. My underlying problem with blanket update package statements is you’re asking the user to do a lot of work that may not have computational benefit and may be disruptive to their other data science projects. You don’t need every single library in your R environment for this OHDSI study. That’s not a good use of anyone’s time. This package is already quite time consuming. Wasting 30+ mins on library installs you don’t even use in this study seems like something we could improve.

**Comment by schuemie** [comment]
What are the 'local environment issues' people are anticipating for 'renv'? Did those people actually try it?

If all you want is a list of the required package versions, then perhaps you're interested in the [insertEnvironmentSnapshotInPackage](http://ohdsi.github.io/OhdsiRTools/reference/insertEnvironmentSnapshotInPackage.html) function in OhdsiRTools. I've been using that function for years to document the versions I had installed at the time I was running the study, as you can see for example [here](https://github.com/OHDSI/StudyProtocols/blob/master/LargeScalePopEst/inst/settings/rEnvironmentSnapshot.csv).

**Comment by gowthamrao** [comment]
Why does it take significant user time to update packages?

**Comment by kmkostka** [comment]
@jdposada mentioned [OS dependencies](https://forums.ohdsi.org/t/using-renv-to-handle-r-package-dependencies/10914/6?u=krfeeney). 

@gowthamrao think of it like this: update.packages crawls your entire R libs and gives you a list of “things to update”. This will include many libraries that have nothing to do with this package’s dependencies. Say there’s 10 libraries with newer versions in CRAN and each takes 2mins to update: that’s 20 mins of updates that have nothing to tangibly do with the study. Why ask a user to do that? It’s not necessary.

Meanwhile, it doesn’t cover the install_github packages that will need to get updated via the loading of the library. So now, a person has run packages updates and is still going to interface with the GitHub dependencies. Sometimes it’s incongruent with the updates they just did so you’re in a circular loop continuously updating small version shifts in what our OHDSI libraries require.

Meanwhile, some machines will throw errors because they have a multiarch install. This seems to be very common in academic environments (Peter reported this, Jason Thomas at UW reported this, Karthik regularly has this issue). It’s not avoidable in environments someone else provisions you resources. So, those people then have the added fun of getting bugs in rJava / install_github, changing where rJava points to and adding a “install-opts” statement. Takes lots of time and persistence. Could easily lead to attrition in who completes the study package.

Bottom line: We should design under the principle to have the lowest common denominator. Asking a user to do anything extra that isn’t related to the dependencies of the package seems superfluous and a waste of their volunteered time. We know that some environments have more problems than others (at no fault of their own). We should not design packages under the assumption that “it’s best practice to update everything” if the package doesn’t require those components and it adds even 20-30 mins of labor. (For those who have errors, it’ll be longer. We see this all the time.) The person time investment in executing a package is a volunteer based effort. We should only ask for what we really need.

**Comment by gowthamrao** [comment]
I see. But in most cases the user time and clock time are different right? I update my packages frequently, but most times it just happens in background while I try to do something else.

**Comment by schuemie** [comment]
We successfully tested renv on Windows, MacOs, and Linux, so I'm not sure what OS dependencies he is referring to. And if someone finds they can't use it, then they can always fall back on the current way of working. (In fact, the renv lock file is still human-readable, as you can see [here](https://github.com/ohdsi-studies/Covid19CohortEvaluation/blob/renv/renv.lock)). Not trying to push anything, I'm just sad that what I think is the exact solution to our problems is being discarded without ever trying.

I would view the multiarch issue on Windows as a separate one that always needs to be solved, regardless of whether you're using renv or not. I don't consider the `install-opts` statement to be a good solution. People should just only install the 64-bit version of R, as recommended [here](https://ohdsi.github.io/Hades/rSetup.html).

**Comment by kmkostka** [comment]
@schuemie Some users do not control the install process and are given an environment by their systems administrator. It’s not their choice what was installed and they can’t easily rectify it. The install-opts is a workaround for when you’re not in a position to install/reinstall in your system and have to deal with the hand that’s dealt to you.

**Comment by schuemie** [comment]
@kmkostka : it looks like renv [always uses `--no-multiarch`](https://github.com/rstudio/renv/blob/master/R/R.R#L165), so could solve this problem as well.

**Comment by anthonysena** [comment]
Closing this issue as we'll aim to use `renv` for a subsequent version of this package to address this discussion.
